Le forum évite de remonter les sujets dont le dernier message est masqué

Rowin opened the pull request:
<!-- Veuillez décrire vos changements à l’emplacement de ce commentaire (inutile dans le cas de tout petits changements). -->
Les fonctions `get_last_post` et `get_last_message` tiennent désormais compte de la visibilité des messages pour ne renvoyer que le dernier message visible.
En particulier : 
-  `forum.models.Topic.get_last_post` ne renvoie plus l'attribute `last_message` mais effectue une requête pour retourner le dernier post visible
- `forum.models.Forum.get_last_message` tient compte de la visibilité dans la requête qui était déjà effectuée.

<!-- Indiquez ci-dessous les numéros des tickets (avec le # au début) corrigés par vos changements : -->
Numéro du ticket concerné (optionnel) : #5980 

<!-- Si certains de vos commits corrigent des bugs, il est préférable de les indiquer également dans les messages de commit en suivant ces instructions : https://help.github.com/articles/closing-issues-using-keywords/ -->

<!-- Si votre pull request nécessite d’effectuer des actions particulières lors de la mise en production, renseignez-les ici afin qu’elles soient ajoutées au changelog lors du merge. -->


### Contrôle qualité

<!-- Écrivez éventuellement ici quelques instructions pour nous aider à vérifier vos changement.

Par exemple :

  - Lancez `python manage.py migrate` et `yarn test` ;
  - Créez un nouveau compte nommé `toto` ;
  - Envoyez un message privé à quelqu’un d’autre, le titre du message doit apparaître en rose. -->

- Se connecter comme staff
- Masquer le dernier message d'un topic
- Sur la vue du forum, le dernier message du topic doit être le message précédent celui qui vient d'être masqué
- Démasquer le message masqué précédemment
- Sur la vue du forum, le dernier message du topic doit à nouveau être le message démasqué

<!-- Merci ! -->
C'est ma première PR ici (et presque ma première de manière générale), n'hésitez pas à me dire si des choses ne vont pas !

entwanne committed:
Ajout d'un filtre par tag sur les flux RSS/Atom des contenus

entwanne committed:
fixup! Ajout d'un filtre par tag sur les flux RSS/Atom des contenus

entwanne committed:
fixup! Ajout d'un filtre par tag sur les flux RSS/Atom des contenus

entwanne committed:
Strip des valeurs des filtres category/subcategory sur les flux RSS/Atom

entwanne committed:
Relaie les paramTres GET de la page courante aux flux RSS

viki53 committed:
Fix hauteur body et z-index header

Rowin committed:
Fix #5980
get_last_post et get_last_message tiennent compte des postes masqués

Rowin committed:
Ajout des tests

Situphen committed:
Correction de l'affichage des membres sur les pages des casquettes

Situphen committed:
Survoler le sous-titre d'une cartouche de sujet du forum ne souligne plus le titre

Situphen committed:
On cache le bouton Modifier dans les MPs si quelqu'un a répondu après

Situphen committed:
Un membre ne peut pas modifier son message s'il est masqué

Situphen committed:
Rend moins visible un paragraphe

AmauryCarrade commented:
Salut, et bienvenue sur le projet ! Merci pour cette contribution qui sera fort utile pour l'ergonomie du forum :) .

Je n'ai pas encore pris le temps de faire une revue extensive, mais je note quelques points auxquels il faudra faire attention. Peut-être que certains sont déjà OK, et alors… tant mieux :p .

1. `last_message` était en cache, donc il faut voir que ça ne pose pas de soucis de performances que de le transformer ainsi.
2. Il faudra vérifier que l'anti-flood (qui accède au dernier message) fonctionne toujours correctement lorsque le dernier message de l'auteur est masqué (pas sûr qu'on ait un test pour ça).
3. Enfin, mais là encore je n'ai pas encore eu le temps de QA, il serait bien de trier les messages par dernier message visible sur la liste des sujets, comme ça les sujets avec un dernier message masqué ne remontent pas non plus dans la liste elle-même.

Si jamais il y a des soucis avec la redéfinition de `last_message()`, le plus simple est probablement de faire une propriété `last_visible_message()` et de l'utiliser là où c'est pertinent (typiquement dans le gabari de la liste des sujets et des forums).

---

P.-S. Si jamais ça peut t'intéresser, on discute aussi du projet de façon plus directe et informelle sur le serveur Discord non-officiel de ZdS.

Situphen committed:
Cache les boutons Spam potentiel et Réponse utile pour les messages masqués

AmauryCarrade reviewed (changes requested):
Bon, du coup j'ai QA. :x: NOK, mais c'est un bon début !

Ce qui fonctionne : 

- tout ce qui est indiqué dans la QA : le bon message est bien affiché sans souci.

Ce qui ne fonctionne pas : 

- l'anti-flood est effectivement cassé : ici, on ne devrait pas pouvoir poster ;
  ![image](https://user-images.githubusercontent.com/1417570/109894722-49c85c80-7c8e-11eb-925e-1409cb8b462f.png)
- l'ordre des messages du forum : il serait mieux, je pense, de garder l'ordre chronologique en ignorant les messages masqués.
  ![image](https://user-images.githubusercontent.com/1417570/109894883-87c58080-7c8e-11eb-85e3-9786fd6bd13c.png)

Faire comme suggéré au précédent commentaire fonctionnerait, je pense, avec une méthode dédiée pour récupérer le dernier message non-masqué, utilisée dans les gabarits des listes de sujets et de forum (car dans la liste des forums, il faudrait aussi afficher le dernier sujet en prenant en compte la date du dernier message non-masqué).

On pourrait envisager de mettre ce message là (non-masqué) en cache également, tout comme le dernier message tout court, ou même à la place, éventuellement, s'il n'y a pas d'usage critique du dernier message tout court.

Ça impliquerait une migration qui enregistrerait le-dit dernier message — n'hésite pas si tu veux t'y coller, on peut aider sans souci si tu débutes avec Django (j'ignore ton niveau exact) :) .

Il faudrait également modifier les vues des listes de sujets afin de trier sur la date du dernier message _visible_, histoire d'éviter d'avoir des ordres chronologiques étranges et de bien répondre à la demande de #5980.

Enfin, vu que ça ne semble pas te faire peur, ajouter un test pour ce cas limite de l'anti-flood (s'assurer que l'envoi est bien bloqué même si le dernier message posté est masqué) serait pertinent, je pense.

Merci encore pour tout ça !

Rowin committed:
Change l'URL de la page de profil en /@pseudo (#6057)

* Fix 6047
Nouvelle url plus simple pour le profil utilisateur
Redirection de l'ancienne url

* Ajout des tests

* Mise à jour du test vérifiant l'url du profil

* Enlever une ligne (commentée) de l'ancien code

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

* Changement du template pour afficher la bonne url

* Changement de l'URL du profile inscrite en dur

Co-authored-by: Situphen <Situphen@users.noreply.github.com>

Situphen committed:
Passage à zmarkdown 9.1.4 pour corriger un bug

Rowin committed:
Création d'une méthode last_visible_post
Ajout d'une propriété "update" pour le Topic

Rowin committed:
Créé un filtre pour trier les topics par date

Rowin committed:
Ajout d'un test de l'anti-spam

Rowin committed:
Ajout d'un test du filtre de tri des topics

Rowin committed:
Merge branch 'dev' into feature-lastposthidden

Rowin committed:
Merge branch 'dev' into feature-lastposthidden

Arnaud-D commented on zds/forum/tests/tests_views.py:
```diff
@@ -1042,6 +1042,24 @@ def test_failure_new_post_stopped_by_anti_spam(self):
+        post_to_hide = PostFactory(topic=topic, author=profile.user, position=topic.last_message.position + 1)
+
+        staff = StaffProfileFactory()
+        self.assertTrue(self.client.login(username=staff.user.username, password="hostel77"))
```
Comme le but est juste de connecter un utilisateur, remplace cette ligne (et toutes celles similaires) par 
```python
self.client.force_login(staff.user)
```
Ça connecte l'utilisateur passé en argument sans avoir besoin du cérémonial avec le mot de passe et la vérification que ça a bien fonctionné.

Rowin replied:
Merci de l'info !

Rowin committed:
Changement dans les tests

Rowin commented:
@AmauryCarrade juste pour info, au cas où tu l'aurai manqué, à priori c'est bon pour moi (ou en tout cas, suffisamment pour que tu repasses un coup de QA je pense)

Situphen committed:
Merge branch 'dev' into feature-lastposthidden

Situphen commented on zds/forum/models.py:
```diff
@@ -228,6 +228,10 @@ def meta_description(self):
         return Topic.__remove_greetings(first_post)[: settings.ZDS_APP["forum"]["description_size"]]
 
+    @property
+    def update(self):
```
`Topic.last_update` me parait plus compréhensible que `Topic.update` mais c'est simplement une suggestion. :)
```suggestion
    def last_update(self):
```

Rowin replied:
Bonne idée, adoptée

Situphen commented on zds/utils/templatetags/topics_sort.py:
```diff
@@ -0,0 +1,13 @@
+    """
+    :return: the topics sorted by last update (last updated first)
+    """
+    return sorted(topics, key=lambda topic: topic.update, reverse=True)
```
Si jamais tu acceptes ma suggestion `Topic.last_update` il faut modifier cette ligne aussi. :)
```suggestion
    return sorted(topics, key=lambda topic: topic.last_update, reverse=True)
```

Situphen commented on zds/forum/models.py:
```diff
@@ -228,6 +228,10 @@ def meta_description(self):
 
+    @property
+    def update(self):
+        return self.get_last_visible_post().update_index_date
```
L'usage de `update_index_date` pose soucis car cette date est par exemple mise à jour lorsqu'un staff masque ou démasque un message. Je suggère d'utiliser `update` s'il y a eu une mise à jour et `pubdate` s'il n'y en a pas eu. Cela évite de détourner `update_index_date` qui est prévu pour l'indexation dans le moteur de recherche seulement. :)
```suggestion
        last_visible_post = self.get_last_visible_post()
        return last_visible_post.update or last_visible_post.pubdate
```

Rowin replied:
Je sentais en l'écrivant que c'était un peu bancal, ta proposition me semble bien. Adoptée !

Situphen reviewed (changes requested):
C'est du très beau travail @Rowin !

Ceci dit étant donné que l'on touche ici aux messages des forums il ne faut vraiment pas faire d'erreurs c'est pourquoi la phase de QA prend du temps ! :)

Rowin committed:
Changement de update pour last_update
Utilisation de Post.update et Post.pubdate
à la place de Post.update_index_date

Rowin committed:
Merge branch 'feature-lastposthidden' of github.com:Rowin/zds-site into feature-lastposthidden

Rowin committed:
Merge branch 'dev' into feature-lastposthidden

Rowin committed:
Linting

Rowin committed:
Merge branch 'dev' into feature-lastposthidden

AmauryCarrade committed:
Merge branch 'dev' into feature-lastposthidden

AmauryCarrade reviewed (dismissed):
:heavy_check_mark: QA OK

Désolé pour le délai, j'ai passé un petit temps sur d'autres choses. Mais j'ai relu le code et testé le comportement et cette fois tout me semble bon : l'ordre est OK, et l'antispam fonctionne correctement. Merci pour cette contribution !

Je ne fusionne pas tout de suite car @Situphen désirait revoir le code lui aussi, mais pour moi c'est ok.

Situphen commented on zds/forum/models.py:
```diff
@@ -232,6 +237,11 @@ def meta_description(self):
 
+    @property
+    def last_update(self):
+        last_visible_post = self.get_last_visible_post()
+        return last_visible_post.update or last_visible_post.pubdate
```
Prenons le cas de deux sujets : 

- Sujet A dont le dernier message est il y a 10 heures ;
- Sujet B dont le dernier message est il y a 11 heures mais modifié il y a 9 heures.

Actuellement en production on affiche Sujet A puis Sujet B, mais sur ta PR on affiche Sujet B puis Sujet A. Pour afficher Sujet A puis Sujet B comme en production, il faut faire ce petit changement : 

```suggestion
        return self.get_last_visible_post().pubdate
```

Rowin replied:
Ok, j'avoue ne pas avoir vérifier en production, le comportement me paraissait logique que la modification fasse remonter le topic (et d'ailleurs, tu te contredis ^^ : https://github.com/zestedesavoir/zds-site/pull/6056#discussion_r614725723).
Je prends note et je corrige !

Rowin replied:
TODO : Rajouter un test pour gérer ce cas spécifiquement. Mon commit n'aurait pas dû casser un comportement existant. Je peux le rajouter dans cette PR, même si ça déborde un peu ?

Situphen reviewed (dismissed):
Je me rends compte que cette PR n'est vraiment pas triviale, il y a encore quelques changements à effectuer !

Actuellement tu tries l'ordre des sujets dans le gabarit avec le *templatetag* `topics_sort` mais ça ne peut pas fonctionner correctement car la liste des sujets est paginée.

Prenons le cas de cinq sujets : 

- Sujet A dont le dernier message est il y a 10 heures ;
- Sujet B dont le dernier message est il y a 11 heures ;
- Sujet C dont le dernier message est il y a 12 heures ;
- Sujet D dont le dernier message est il y a 13 heures ;
- Sujet E dont le dernier message est il y a 14 heures.

Si la pagination est configurée pour n'afficher que 3 sujets par page, les sujets affichés sur la première page sont dans l'ordre Sujet E, Sujet D et Sujet C.

Si le dernier message de Sujet B est édité, les sujets affichés sur la première page sont dans l'ordre Sujet B, Sujet E et Sujet D.

Si ensuite le dernier message de Sujet est masqué, on s'attend à ce que les sujets affichés sur la première page redeviennent dans l'ordre Sujet E, Sujet D et Sujet C. Or ce n'est pas le cas actuellement avec ta PR, les sujets affichés sur la première page sont Sujet E, Sujet D, Sujet B.

Étant donné que la page est paginée, la liste `topics` dans `templates/forum/category/forum.html` ne contient que dans l'ordre Sujet B, Sujet E et Sujet D, ce que fait que `topics|topics_sort` contient dans l'ordre Sujet E, Sujet D, Sujet B.

Si tu veux reproduire cela chez toi, il faut modifier `topics_per_page` dans `zds/abstract_base/zds.py` pour modifier le nombre de sujets affichés par page.

Il faudrait donc que tu fasses ce tri avant la pagination. Les requêtes sont : 

- pour la liste des sujets épinglés
  https://github.com/zestedesavoir/zds-site/blob/415038cd12df9d57bfb16d58d4d8d39f2a90126a/zds/forum/views.py#L124
- pour le reste des sujets
  https://github.com/zestedesavoir/zds-site/blob/415038cd12df9d57bfb16d58d4d8d39f2a90126a/zds/forum/views.py#L158

Le mieux est donc probablement que tu modifies ceci : 

https://github.com/zestedesavoir/zds-site/blob/415038cd12df9d57bfb16d58d4d8d39f2a90126a/zds/forum/managers.py#L93-L100

Il faudra voir si les changements posent des soucis de performance ou pas (notamment regarder si le nombre de requêtes SQL augmente ou pas).

Rowin committed:
Corrige un problème d'ordre des derniers sujets lorsque les sujets sont paginés

Rowin committed:
Corrige un problème d'ordre des sujets lorsqu'un message a été modifié

Lorsque le dernier message visible était modifié, il était remonté. Ce n'est pas le comportement attendu. Correction de ce comportement.

Rowin commented:
@Situphen je pense avoir trouvé une solution propre pour gérer ce cas, directement dans le manager et sans rajouter de requêtes en rajoutant une annotation qui contient la date. Ça m'évite d'avoir à gérer une persistance ou a rajouter des requêtes supplémentaires !

Rowin committed:
Merge branch 'dev' into feature-lastposthidden

Rowin committed:
Merge branch 'dev' into feature-lastposthidden

Arnaud-D commented:
@Situphen tu auras le temps de repasser ici ? C'était toi qui avait fait la QA à l'époque et j'ai l'impression que c'est presque bon. Sauf peut-être l'histoire de performance, je ne sais pas si c'est facile à QAiser.

Situphen commented:
> @Situphen tu auras le temps de repasser ici ? C'était toi qui avait fait la QA à l'époque et j'ai l'impression que c'est presque bon. Sauf peut-être l'histoire de performance, je ne sais pas si c'est facile à QAiser.

Je t'avoue qu'en ce moment le développement technique de ZDS ne m'intéresse pas trop, donc il est peu probable que je finisse la QA de cette PR. Libre à toi ou quelqu'un d'autre d'en faire la QA ! Je ne met aucun veto à ce qu'elle soit fusionné si quelqu'un juge que c'est bon :)

philippemilink commented on templates/forum/category/forum.html:
```diff
@@ -86,7 +87,7 @@
     {% if topics %}
         <div class="topic-list navigable-list" itemscope itemtype="http://schema.org/ItemList">
-            {% for topic in topics %}
+            {% for topic in topics|topics_sort %}
```
Puisque le tri est fait dans `get_all_topics_of_a_forum()`, on n'a plus besoin de trier ici, non ?

philippemilink commented on zds/forum/models.py:
```diff
@@ -232,6 +237,11 @@ def meta_description(self):
 
+    @property
+    def last_update(self):
+        last_visible_post = self.get_last_visible_post()
+        return last_visible_post.pubdate
```
`self.get_last_visible_post()` peut retourner `None`, dans ce cas, on ne peut pas accéder à l'attribut `pubdate` et ça va générer une erreur.

philippemilink reviewed (changes requested):
Deux petites questions sur le code, et il faudrait régler le conflit sur le fichier `zds/utils/tests/test_misc.py`.

Une fois que ces modifications seront faites, je testerai, ça me semble bien.

PS : @Rowin puisque cette PR est ouverte depuis longtemps, dis-le si tu n'as pas le temps / l'envie / la motivation / ... de faire les modifications demandées, pour qu'on sache de notre côté comment considérer cette PR :)
